### DISCUSSION TITLE
STM32RTC with Arduino Portenta H7 based on ArduinoCore-mbed

### DISCUSSION THREAD

**Discussion by hideakitai** [stm32duino/Arduino_Core_STM32#2336] [Libraries]
Hi, thanks for the great work! I forked the [STM32RTC](https://github.com/stm32duino/STM32RTC) library because I'd like to make it work with Arduino Portenta H7 (STM32H747XIH6) based on [ArduinoCore-mbed](https://github.com/arduino/ArduinoCore-mbed).

https://github.com/hideakitai/STM32RTCMbed

I built it using ArduinoCore-mbed and got most of the basic RTC functionality examples working.

https://github.com/hideakitai/STM32RTCMbed/pull/1

However, only the Alarm functionality is not working. It looks like `RTC_Alarm_IRQHandler()` is not being called. So, let me ask two questions:

1. If you use this library with Arduino Portenta H7 (STM32H747XIH6) based on Arduino_Core_STM32, can you get the Alarm to work successfully? (Sorry, I don't have ST-Link, so I haven't confirmed if it works with Arduino_Core_STM32 and STM32RTC...) 

2. If no.1 works fine, do you have any idea what the cause might be, based on your abundant and deep experience with STM32 HAL/LL? Any information or advice would be greatly appreciated (I'm not very familiar with STM32 HAL/LL. The code looks good to me, but perhaps my code or something in ArduinoCore-mbed or mbed-os has some problems).

I will briefly explain what I did in the above pull request to get STM32RTC to work with ArduinoCore-mbed.

- [b3506cd](https://github.com/hideakitai/STM32RTCMbed/pull/1/commits/b3506cda88a140ba37a6a833cbb0b9db8bb5fd79) First, I imported codes from Arduino_Core_STM32, which were included from STM32RTC but did not exist in ArduinoCore-mbed.
- [4146b46](https://github.com/hideakitai/STM32RTCMbed/pull/1/commits/4146b467cdb9bd6fd402813749629c5c2d346001) Secondly, no macro is defined to determine the series, such as `STM32H7xx`, so I defined one.
- [d6d09e5](https://github.com/hideakitai/STM32RTCMbed/pull/1/commits/d6d09e5d991dc2e47beae006bf3d74f4d1dde0a1) I renamed `voidFuncPtr` to avoid conflict with the ArduinoCore-mbed definition.
- [ed28495](https://github.com/hideakitai/STM32RTCMbed/pull/1/commits/ed2849594a6eb4a776b1f4129fb17382bced6751) Changed `SysTick_Handler()` to `WEAK` since it seems to be already implemented in mbed-os (`libmbed.a`) and causes a "multiple definitions" linker error.

That's all. As mentioned above, the basic RTC example works fine. Only the Alarm Interrupt is not working. 

I would greatly appreciate your experience and knowledge. Thank you very much for your valuable time.

**Comment by fpistm** [comment]
Hi @hideakitai 

> * If you use this library with Arduino Portenta H7 (STM32H747XIH6) based on Arduino_Core_STM32, can you get the Alarm to work successfully? (Sorry, I don't have ST-Link, so I haven't confirmed if it works with Arduino_Core_STM32 and STM32RTC...)
> *

Never tested as the core doesn't support the Arduino Portenta board.

About the second point, I guess the libmed.a already implement the `RTC_Alarm_IRQHandler()`. So your one is simply not called.

**Reply by hideakitai** [reply]
> Never tested as the core doesn't support the Arduino Portenta board.

Ah, added to the list but not implemented yet? I got it https://github.com/stm32duino/Arduino_Core_STM32/issues/722

> About the second point, I guess the libmed.a already implement the RTC_Alarm_IRQHandler(). So your one is simply not called.

Thank you for the great point! I will try to override the implementation by `--wrap`ping `RTC_Alarm_IRQHandler()`.

**Reply by hideakitai** [reply]
Hmm... not working. I will explore more!

**Reply by hideakitai** [reply]
In my next experiment with STM32LowPowerMbed (STM32LowPower for ArduionoCore-mbed), the RTC alarm seems to be enabled because `STM32LowPower::setAlarmTime(ms)` works correctly. 

https://github.com/hideakitai/STM32LowPowerMbed/blob/b05838f9309089dfc5b5bbb40173ec0f82f11dc2/src/STM32LowPowerMbed.cpp#L231

In this function, the RTC alarm is set and enabled like:

https://github.com/hideakitai/STM32LowPowerMbed/blob/b05838f9309089dfc5b5bbb40173ec0f82f11dc2/src/STM32LowPowerMbed.cpp#L318-L320

So only the RTC alarm IRQ handler is not working, maybe...

**Reply by hideakitai** [reply]
I was mistaken. It seems that the examples related to RTC Alarm is not working after all...

**Reply by fpistm** [reply]
Well hard to help on this. You should probably ask on Arduino forum or ArduinoCore-mbed GitHub.
You could also try to debug and see if the Alarm is properly configured and if the ALARMA IT is raised.

**Reply by hideakitai** [reply]
Thanks for your support! I will debug while familiarizing myself with how to handle HAL. If you notice any STM32H7-specific settings, please let me know. (I think it is probably due to the override of HAL functions on the ArduinoCore-mbed or mbed-os side)
